Remove caffe2::ShareData

jerryzh168 opened the pull request:
Summary:
Previously we are using Resize + ShareData.
Instead, we'll create a function on Tensor that clones itself with same storage.

Suppose we want `t` to `ShareData` with `t0`, Previous:
```
Tensor t(dims, CPU);
t.Resize(t0.sizes());
t.ShareData(t0);
```
Now:
```
Tensor t = t0.Alias();
```

Differential Revision: D13507609

bddppq commented:
@pytorchbot retest this please

jerryzh168 commented:
@pytorchbot retest this please

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -102,6 +102,25 @@ class CAFFE2_API Tensor final {
   }
 
+  /**
+   * Clone self as a Tensor that share the same Storage
```
I'd like this doc to be beefier. What does it mean to "share a same storage"? If I subsequently resize one of the aliases of the tensor, what happens? What if I restride it in-place?

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -102,6 +102,25 @@ class CAFFE2_API Tensor final {
+        storage_initialized(),
+        "Cloning a tensor that has no content and has size > 0");
+    x.impl_->set_storage(storage());
+    x.raw_mutable_data(dtype());
```
Does this do the right thing when dtype is not initialized? I'm not convinced...

jerryzh168 replied:
Looking again, we probably don't need to call raw_mutable_data, since set_storage already set data_type_ of TensorImpl to be that of Storage.

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -102,6 +102,25 @@ class CAFFE2_API Tensor final {
+  /**
+   * Clone self as a Tensor that share the same Storage
+   */
+  Tensor Alias() const {
```
TODO: Deduplicate this with `THTensor_(newWithTensor)` (exposed in ATen as at::alias but not otherwise available)

```
/* Pointer-copy init */
THTensor *THTensor_(newWithTensor)(THTensor *tensor)
{
  THTensor *self = c10::make_intrusive<at::TensorImpl, at::UndefinedTensorImpl>(
    c10::intrusive_ptr<at::StorageImpl>::reclaim(THStorage_(new)()),
    at::CPUTensorId(),
    false
  ).release();
  THTensor_(setStorageNd)(self,
                          THTensor_getStoragePtr(tensor),
                          tensor->storage_offset(),
                          tensor->dim(),
                          THTensor_getSizePtr(tensor),
                          THTensor_getStridePtr(tensor));
  return self;
}
```

jerryzh168 replied:
you mean add this TODO in the comment right?

jerryzh168 replied:
probably better to do it in another diff?

ezyang commented on aten/src/ATen/test/tensor_interop_test.cpp:
```diff
@@ -66,11 +66,11 @@ TEST(TestTensorInterop, PytorchToCaffe2Op) {
-    c2_tensor_c.ResizeLike(at_tensor_c.getIntrusivePtr());
-    c2_tensor_c.ShareData(at_tensor_c.getIntrusivePtr());
+    caffe2::Tensor c2_tensor_from_aten(at_tensor_c.getIntrusivePtr());
+    BlobSetTensor(workspace.CreateBlob("c"), c2_tensor_from_aten.Alias());
```
This test seems woefully inadequate for actually checking if aliasing has occurred, but maybe that is not the point of the test.

jerryzh168 replied:
yeah..

ezyang commented on caffe2/operators/softmax_ops.cu:
```diff
@@ -482,20 +482,21 @@ bool SoftmaxWithLossGradientOp<float, CUDAContext>::RunOnDevice() {
+  Tensor* dX;
+  if (only_loss_) {
+    // Memory saving trick to share the buffer with the softmax output.
+    // Softmax output is thus overwritten.
```
You have done nothing wrong here, but this code makes my skin crawl...

ezyang commented:
The code changes look legit; I just want beefier docs.

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 committed:
Remove caffe2::ShareData (#15418)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15418

Previously we are using Resize + ShareData.
Instead, we'll create a function on Tensor that clones itself with same storage.

Suppose we want `t` to `ShareData` with `t0`, Previous:
```
Tensor t(dims, CPU);
t.Resize(t0.sizes());
t.ShareData(t0);
```
Now:
```
Tensor t = t0.Alias();
```

Reviewed By: dzhulgakov

Differential Revision: D13507609

fbshipit-source-id: c99b76242fee22063d8ac44fd6f5df3bf6006551
